tensorflow/core/kernels/stack.cc

```diff
@@ -72,19 +80,95 @@ class Stack : public ResourceBase {
       return errors::InvalidArgument("Stack[", stack_name_,
                                      "] is empty when calling Pop().");
     }
-    *value = stack_.back();
+    int index = stack_.size() - 1;
+    while (is_swapping_ins_[index]) {
+      cond_swapping_ins_[index].wait(l);
+      index = stack_.size() - 1;
+    }
+    *value = stack_.back().first;
+    absl::optional<std::list<int>::iterator> iter = stack_.back().second;
+    if (iter) {
+      unswapped_lru_.erase(*iter);
+    }
     stack_.pop_back();
     return Status::OK();
   }
 
+  Status MarkAsSwapped(TensorAndAllocation* value, Tensor* cpu_tensor) {
+    mutex_lock l(mu_);
+    TF_RETURN_IF_ERROR(CheckNotClosed());
+
+    value->tensor = *cpu_tensor;
+    value->swapped_to_cpu = true;
+
+    return Status::OK();
+  }
+
+  Status MarkAsUnswapped(TensorAndAllocation* value, Tensor* device_tensor) {
+    mutex_lock l(mu_);
+    TF_RETURN_IF_ERROR(CheckNotClosed());
+
+    value->tensor = *device_tensor;
+    value->swapped_to_cpu = false;
+
+    return Status::OK();
+  }
+
+  bool GetTensorToSwapOut(std::function<bool(const TensorAndAllocation&)> cond,
+                          TensorAndAllocation** value) {
+    mutex_lock l(mu_);
+    bool can_swap = false;
+    while (!can_swap) {
+      if (unswapped_lru_.empty()) {
+        break;
+      }
+      int index = unswapped_lru_.front();
+      unswapped_lru_.pop_front();
+      if (index < stack_.size()) {
```

Comment:
We shouldn't need this if statement anymore if the data structures are always consistent with one another.

Reply:
You're right, I forgot about this, removed.